In [148]:
import pandas as pd

In [149]:
df = pd.read_csv("diabetes.csv")

In [150]:
df.head()

col2  col2.1  col3  col4  col5  col6   col7  col8  y
0     6     148    72    35     0  33.6  0.627    50  1
1     1      85    66    29     0  26.6  0.351    31  0
2     8     183    64     0     0  23.3  0.672    32  1
3     1      89    66    23    94  28.1  0.167    21  0
4     0     137    40    35   168  43.1  2.288    33  1

In [151]:
df.describe()

col2      col2.1        col3        col4        col5        col6  \
count  768.000000  768.000000  768.000000  768.000000  768.000000  768.000000   
mean     3.845052  120.894531   69.105469   20.536458   79.799479   31.992578   
std      3.369578   31.972618   19.355807   15.952218  115.244002    7.884160   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      1.000000   99.000000   62.000000    0.000000    0.000000   27.300000   
50%      3.000000  117.000000   72.000000   23.000000   30.500000   32.000000   
75%      6.000000  140.250000   80.000000   32.000000  127.250000   36.600000   
max     17.000000  199.000000  122.000000   99.000000  846.000000   67.100000   

             col7        col8           y  
count  768.000000  768.000000  768.000000  
mean     0.471876   33.240885    0.348958  
std      0.331329   11.760232    0.476951  
min      0.078000   21.000000    0.000000  
25%      0.243750   24.000000    0.000000  
50%      0.372500   29.000000    0.000000  
75%      0.626250   41.000000    1.000000  
max      2.420000   81.000000    1.000000

SPLITING THE DATASET INTO TRAIN AND TEST



In [152]:
from sklearn.model_selection import train_test_split

In [153]:
X, y = df.iloc[:, :8].values, df.iloc[:, 8].values
X_train, X_test, y_train, y_test = \
    train_test_split(X, y, test_size=0.3, random_state=0)

# **NORMALIZE THE DATASET**

---



In [154]:
from sklearn.preprocessing import MinMaxScaler
mms = MinMaxScaler()
X_train_norm = mms.fit_transform(X_train)
X_test_norm = mms.transform(X_test)

# **APPLYING PERCEPTRON ALGOIRTHM**

---



In [155]:
from sklearn.linear_model import Perceptron
model = Perceptron(random_state=0 , max_iter=20, eta0=0.001)
model.fit(X_train_norm, y_train)



Perceptron(eta0=0.001, max_iter=20)

In [156]:
from sklearn.metrics import accuracy_score
predictions_train = model.predict(X_train_norm)
predictions_test = model.predict(X_test_norm)
train_score = accuracy_score(predictions_train, y_train)
print("score on train data: ", train_score)
test_score = accuracy_score(predictions_test, y_test)
print("score on test data: ", test_score)



score on train data:  0.7318435754189944
score on test data:  0.7316017316017316


In [157]:
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(y_test, predictions_test)
confusion_matrix


array([[153,   4],
       [ 58,  16]])

In [158]:
from sklearn.metrics import classification_report
print(classification_report(y_test, predictions_test))

              precision    recall  f1-score   support

           0       0.73      0.97      0.83       157
           1       0.80      0.22      0.34        74

    accuracy                           0.73       231
   macro avg       0.76      0.60      0.59       231
weighted avg       0.75      0.73      0.67       231



# **APPLYING ADALINE ALGOIRTHM**

---



In [159]:
import numpy as np

class AdalineGD(object):
    """ADAptive LInear NEuron classifier.

    Parameters
    ------------
    eta : float
        Learning rate (between 0.0 and 1.0)
    n_iter : int
        Passes over the training dataset.

    Attributes
    -----------
    w_ : 1d-array
        Weights after fitting.
    cost_ : list
        Sum-of-squares cost function value in each epoch.

    """
    def __init__(self, eta=0.01, n_iter=50):
        self.eta = eta
        self.n_iter = n_iter

    def fit(self, X, y):
        """ Fit training data.

        Parameters
        ----------
        X : {array-like}, shape = [n_samples, n_features]
            Training vectors, where n_samples is the number of samples and
            n_features is the number of features.
        y : array-like, shape = [n_samples]
            Target values.

        Returns
        -------
        self : object

        """
        self.w_ = np.zeros(1 + X.shape[1])
        self.cost_ = []

        for i in range(self.n_iter):
            net_input = self.net_input(X)
            # Please note that the "activation" method has no effect
            # in the code since it is simply an identity function. We
            # could write `output = self.net_input(X)` directly instead.
            # The purpose of the activation is more conceptual, i.e.,  
            # in the case of logistic regression, we could change it to
            # a sigmoid function to implement a logistic regression classifier.
            output = self.activation(X)
            errors = (y - output)
            self.w_[1:] += self.eta * X.T.dot(errors)
            self.w_[0] += self.eta * errors.sum()
            cost = (errors**2).sum() / 2.0
            self.cost_.append(cost)
        return self

    def net_input(self, X):
        """Calculate net input"""
        return np.dot(X, self.w_[1:]) + self.w_[0]

    def activation(self, X):
        """Compute linear activation"""
        return self.net_input(X)

    def predict(self, X):
        """Return class label after unit step"""
        return np.where(self.activation(X) >= 0.0, 1, -1)

In [160]:
ada1 = AdalineGD(n_iter=20, eta=0.001).fit(X_train_norm, y_train)

In [161]:
from sklearn.metrics import accuracy_score
predictions_train = ada1.predict(X_train_norm)
predictions_test = ada1.predict(X_test_norm)
train_score = accuracy_score(predictions_train, y_train)
print("score on train data: ", train_score)
test_score = accuracy_score(predictions_test, y_test)
print("score on test data: ", test_score)


score on train data:  0.3612662942271881
score on test data:  0.3203463203463203


In [162]:
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(y_test, predictions_test)
confusion_matrix


array([[  0, 157],
       [  0,  74]])

In [163]:
from sklearn.metrics import classification_report
print(classification_report(y_test, predictions_test))

              precision    recall  f1-score   support

           0       0.00      0.00      0.00       157
           1       0.32      1.00      0.49        74

    accuracy                           0.32       231
   macro avg       0.16      0.50      0.24       231
weighted avg       0.10      0.32      0.16       231



/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# **APPLYING Logisic Regression ALGOIRTHM**

---



In [164]:
from sklearn.linear_model import LogisticRegression
model3 = LogisticRegression(random_state=0 , max_iter = 20).fit(X_train_norm, y_train)



In [165]:
from sklearn.metrics import accuracy_score
predictions_train = model3.predict(X_train_norm)
predictions_test = model3.predict(X_test_norm)
train_score = accuracy_score(predictions_train, y_train)
print("score on train data: ", train_score)
test_score = accuracy_score(predictions_test, y_test)
print("score on test data: ", test_score)


score on train data:  0.7579143389199255
score on test data:  0.7662337662337663


In [166]:
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(y_test, predictions_test)
confusion_matrix


array([[143,  14],
       [ 40,  34]])

In [167]:
from sklearn.metrics import classification_report
print(classification_report(y_test, predictions_test))

              precision    recall  f1-score   support

           0       0.78      0.91      0.84       157
           1       0.71      0.46      0.56        74

    accuracy                           0.77       231
   macro avg       0.74      0.69      0.70       231
weighted avg       0.76      0.77      0.75       231



# ****

---



# **Which algorithm gives best accuracy? **

---



Perceptron:
score on train data:  0.7318435754189944,
score on test data:  0.7316017316017316


Adaline:
score on train data:  0.3612662942271881,
score on test data:  0.3203463203463203

Logistic Regression:
score on train data:  0.7579143389199255
score on test data:  0.7662337662337663



# ** Confusion Matrix on all three ALGOIRTHMs**

---



Perceptron: 
      
      array([[153,   4],

       [ 58,  16]])

Adaline: 

      array([[  0, 157],

       [  0,  74]])

Logistic Regression: 

      array([[143,  14],

       [ 40,  34]])